### PR TITLE
Comments for curves and pairing primitives

### PR DESCRIPTION
**Pull request by UlrichHaboeck75**
Comments/References for all implemented curves, including the non-pairing friendly ones.
General description of Ate pairing for BLS12, without details on pre-computations.

Update: Run ``cargo doc`` and browse through ``target/doc/algebra/curves/`` to see how the cargo documentation looks like.

### CODE REVIEW

File: algebra/src/curves/bls12_377/mod.rs [hunk 1]
```diff
@@ -1,3 +1,15 @@
...
+//! inner curve of the sw6.
+//! Its a BLS curve with embedding degree 12 and base field size is 377 bits.
+//!
+//! Its security against modern version of the number field sieve (STNFS) is estimated as 125 bit
+//! by [HG 2020](https://eprint.iacr.org/2020/351.pdf).
+
+// The HG 2020 result states an unexpected negligible loss in security compared to the recommendations
+// from [Guillevic 2019](https://eprint.iacr.org/2019/1371) which establishes a base field size
+// of 446 Bit to accomodate that security level.
+// I hope that HG will give us some hints on that.
```
Comment (mauriziobinello) [inline algebra/src/curves/bls12_377/mod.rs:11]: I would remove the personal flavor in the comment "I hope that".  And I wouldn't use in any case the word "hope" here. We can make it a note for future check or future action.

File: algebra/src/curves/mnt4753/mod.rs [hunk 2]
```diff
@@ -1,3 +1,11 @@
+//! [Coda's MNT4-753](https://codaprotocol.github.io/snark-challenge/MNT4753.html) curve of
+//! its MNT4/6 cycle for fully recursive SNARKs.
+//! Both base field as well as prime order subgroup are of size 753 bits.
+//!
+//! Supposed to satisfy a security level of 128 Bit, but a recent estimate in the manner of
```
Comment (mauriziobinello) [inline algebra/src/curves/mnt4753/mod.rs:5]: "a recent estimate", assumes the code is tagged with a date. I would replace with the date of the paper. Also, in general, we are calculating the security level of the various curves - maybe we should add in all of those comments also a date there, e.g. as of June 2020, it serves a security level of ... bits.

**Review by mauriziobinello — COMMENTED** [review]
I added some notes in the added comments

File: algebra/src/curves/mnt4753/mod.rs (hunk 2 above)
Comment (UlrichHaboeck75) [reply algebra/src/curves/mnt4753/mod.rs:5]: The "recent estimate" refers to Aurore's note that she has sent us, but is not published anywhere. Maybe we should put a link to the file on ginger (but I am afraid, that this won't be a persistent one).

File: algebra/src/curves/bls12_377/mod.rs (hunk 1 above)
Comment (UlrichHaboeck75) [reply algebra/src/curves/bls12_377/mod.rs:11]: Yes, I agree. The "I hope" was referring to the feedback of Y. Housni, one of the authors of [GH 2020](https://eprint.iacr.org/2020/351.pdf), in the corresponding [zexe issue](https://github.com/scipr-lab/zexe/issues/206). He pointed out the security estimate of the BLS12-377 in that paper, but to my modest understanding, this estimate seems not to go together with the recommendations in [Guillevic 2019](https://eprint.iacr.org/2019/1371).

**Review by 95DDB — APPROVED** [review]

**Pull request merged**